File: outputs/MiniSeries_Studio/automation/lib/elevenlabs.js
// ElevenLabs Flash v2.5 streaming TTS wrapper.
// Voices are keyed by character (Maya, Ella, Kiri, etc.) — IDs come from .env.

import axios from "axios";
import fs from "node:fs";
import path from "node:path";

const API = "https://api.elevenlabs.io/v1";
const MODEL = process.env.ELEVENLABS_MODEL || "eleven_flash_v2_5";

const VOICES = {
  maya: process.env.ELEVENLABS_VOICE_MAYA,
  future_maya: process.env.ELEVENLABS_VOICE_FUTURE_MAYA,
  ella: process.env.ELEVENLABS_VOICE_ELLA,
  kiri: process.env.ELEVENLABS_VOICE_KIRI,
  narrator_s4: process.env.ELEVENLABS_VOICE_NARRATOR_S4,
};

const DEFAULT_SETTINGS = {
  stability: 0.45,
  similarity_boost: 0.75,
  style: 0.2,
  use_speaker_boost: true,
};

/**
 * Render a single VO line to MP3. `character` must match VOICES keys above.
 * Returns the absolute file path of the saved MP3.
 */
export async function tts({ character, text, outFile, settings = {} }) {
  const voiceId = VOICES[character];
  if (!voiceId) throw new Error(`No voice configured for "${character}"`);
  const url = `${API}/text-to-speech/${voiceId}/stream?optimize_streaming_latency=2`;
  const body = {
    text,
    model_id: MODEL,
    voice_settings: { ...DEFAULT_SETTINGS, ...settings },
  };
  const res = await axios.post(url, body, {
    headers: {
      "xi-api-key": process.env.ELEVENLABS_API_KEY,
      Accept: "audio/mpeg",
      "Content-Type": "application/json",
    },
    responseType: "stream",
    timeout: 60_000,
  });
  await fs.promises.mkdir(path.dirname(outFile), { recursive: true });
  await new Promise((resolve, reject) => {
    const w = fs.createWriteStream(outFile);
    res.data.pipe(w);
    w.on("finish", resolve);
    w.on("error", reject);
  });
  return outFile;
}

/** Batch TTS — accepts an array of {character, text, outFile} and runs sequentially. */
export async function ttsBatch(lines) {
  const results = [];
  for (const line of lines) results.push(await tts(line));
  return results;
}

/** Estimate character count for guard-rail accounting. */
export function charsIn(lines) {
  return lines.reduce((sum, l) => sum + (l.text?.length || 0), 0);
}

File: outputs/MiniSeries_Studio/mcp_server/vertical_studio_mcp/elevenlabs.py
"""ElevenLabs client — text-to-speech and voice listing.

API docs: https://elevenlabs.io/docs/api-reference
Endpoints used:
  POST /v1/text-to-speech/{voice_id}     — synthesize speech (returns mp3 bytes)
  GET  /v1/voices                        — list voices in the account
"""
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Optional

import httpx

from . import credentials as cred

BASE = "https://api.elevenlabs.io/v1"


class ElevenLabsError(RuntimeError):
    pass


def _headers(json: bool = False) -> dict:
    key = cred.get_secret("ELEVENLABS_API_KEY")
    if not key:
        raise ElevenLabsError("ELEVENLABS_API_KEY not configured. Run: vertical-studio-keys set ELEVENLABS_API_KEY")
    h = {"xi-api-key": key}
    if json:
        h["Content-Type"] = "application/json"
    return h


@dataclass
class Voice:
    voice_id: str
    name: str
    labels: dict
    preview_url: Optional[str]


def list_voices() -> list[Voice]:
    with httpx.Client(timeout=30) as c:
        r = c.get(f"{BASE}/voices", headers=_headers())
        if r.status_code >= 400:
            raise ElevenLabsError(f"voices {r.status_code}: {r.text}")
        out = []
        for v in r.json().get("voices", []):
            out.append(Voice(
                voice_id=v["voice_id"],
                name=v.get("name", ""),
                labels=v.get("labels", {}) or {},
                preview_url=v.get("preview_url"),
            ))
        return out


def synthesize(
    *,
    text: str,
    voice_id: str,
    dest: Path,
    model_id: str = "eleven_multilingual_v2",
    stability: float = 0.45,
    similarity_boost: float = 0.75,
    style: float = 0.30,
    use_speaker_boost: bool = True,
) -> Path:
    """Render `text` with `voice_id` to an mp3 at `dest`."""
    body = {
        "text": text,
        "model_id": model_id,
        "voice_settings": {
            "stability": stability,
            "similarity_boost": similarity_boost,
            "style": style,
            "use_speaker_boost": use_speaker_boost,
        },
    }
    dest.parent.mkdir(parents=True, exist_ok=True)
    with httpx.Client(timeout=180) as c:
        r = c.post(
            f"{BASE}/text-to-speech/{voice_id}",
            headers=_headers(json=True),
            json=body,
        )
        if r.status_code >= 400:
            raise ElevenLabsError(f"tts {r.status_code}: {r.text}")
        dest.write_bytes(r.content)
    return dest
